src/adapter/rust_type_name.rs

```diff
@@ -0,0 +1,676 @@
+
+            enum Arg<'a> {
+                Named(&'a str, Type<'a>),
+                Dots,
+            }
+
+            intersperse(
+                f,
+                ", ",
+                fnp.sig
+                    .inputs
+                    .iter()
+                    .map(|(name, ty)| Arg::Named(name, Type(ty)))
+                    .chain(fnp.sig.is_c_variadic.then_some(Arg::Dots)),
+                |arg, f| match arg {
+                    Arg::Named(name, ty) => write!(f, "{name}: {ty}"),
+                    Arg::Dots => write!(f, "..."),
+                },
+            )?;
+
+            write!(f, ")")?;
+
+            if let Some(output) = &fnp.sig.output {
+                write!(f, " -> {}", Type(output))?;
+            }
+
+            Ok(())
+        }
+        rustdoc_types::Type::Tuple(tys) => {
+            write!(f, "(")?;
+            intersperse_with(f, ", ", tys, Type)?;
+            write!(f, ")")
+        }
+        rustdoc_types::Type::Slice(ty) => write!(f, "[{}]", Type(ty)),
+        // According to the docs for this variant, `len` is not guaranteed to be the
+        // same as the source code.  For example, [u8; 1 + 2] currently becomes [u8; 3]
+        rustdoc_types::Type::Array { type_, len } => write!(f, "[{}; {}]", Type(type_), len),
+        rustdoc_types::Type::ImplTrait(vec) => {
+            write!(f, "impl ")?;
+            intersperse_with(f, " + ", vec, GenericBound)
+        }
+        rustdoc_types::Type::Infer => write!(f, "_"),
+        rustdoc_types::Type::RawPointer { is_mutable, type_ } => {
+            let kind = if *is_mutable { "mut" } else { "const" };
+
+            write!(f, "*{kind} {}", Type(type_))
+        }
+        rustdoc_types::Type::BorrowedRef {
+            lifetime,
+            is_mutable,
+            type_,
+        } => {
+            write!(f, "&")?;
+            if let Some(lt) = lifetime {
+                write!(f, "{lt} ")?;
+            }
+
+            if *is_mutable {
+                write!(f, "mut ")?;
+            }
+
+            // References can create ambiguity when there is an `impl/dyn Trait + ...`
+            // block:
+            //
+            // `&dyn Read + Send + Sync` is rejected by the compiler, it should be
+            // `&(dyn Read + Send + Sync)`.
+            //
+            // Parentheses for an `impl/dyn Trait` with no `+` are not always required,
+            // but it can create ambiguity with e.g., `impl Fn() -> &dyn Trait + Send + Sync`,
+            // which could be `Fn() -> &(dyn Trait) + Send + Sync`
+            // or `Fn() -> &(dyn Trait + Send + Sync)`.
+            //
+            // Unconditionally wrapping an `impl/dyn Trait` reference with parentheses
+            // may create something syntactially different from the original statement,
+            // but it will always be semantically equivalent, and lets the formatting be more
+            // context-free than keeping track of when parentheses are needed.
```

Comment:
Would this produce `&(dyn Trait)` notation though?

The eventual use of this code would be in materializing witness code, which we expect users will scrutinize and nitpick on. I'm not sure how much work it would be to only emit parens when strictly necessary, but it might be really good to look into it (possibly in a future PR) so that our generated Rust code looks as idiomatic as possible.

People will mistrust witnesses that they can't read and understand, so this is one place where paying a bit more development cost one time up front will probably pay off in the long run: many if not most of our witness generators will rely on this code.